federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,827 @@
+	//                 2. need one kubeClient per cluster, or look up based on type of cluster.
+	//                 3. Replace clustername with cluster api object reference.
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(&unversioned_core.EventSinkImpl{Interface: kubeClient.Core().Events("")})
+	recorder := broadcaster.NewRecorder(api.EventSource{Component: "federated-service-controller"})
+
+	return &ServiceController{
+		cloud:            cloud,
+		kubeClient:       kubeClient,
+		clusterName:      clusterName,
+		cache:            &serviceCache{serviceMap: make(map[string]*cachedService)},
+		eventBroadcaster: broadcaster,
+		eventRecorder:    recorder,
+		// TODO: quinton: We watch for changes to underlying Kubernetes services, not notes.  Update this accordingly.
+		nodeLister: cache.StoreToNodeLister{
+			Store: cache.NewStore(cache.MetaNamespaceKeyFunc),
+		},
+	}
+}
+
+// Run starts a background goroutine that watches for changes to Federated services
+// and ensures that they have Kubernetes services created, updated or deleted appropriately.
+// federationSyncPeriod controls how often we check the federation's services to
+// ensure that the correct Kubernetes services (and associated DNS entries) exist.
+// This is only necessary to fudge over failed watches.
+// clusterSyncPeriod controls how often we check the federation's underlying clusters and
+// their Kubernetes services to ensure that matching services created independently of the Federation
+// (e.g. directly via the underlying cluster's API) are correctly accounted for.
+
+// It's an error to call Run() more than once for a given ServiceController
+// object.
+func (s *ServiceController) Run(federationSyncPeriod, clusterSyncPeriod time.Duration) error {
+	if err := s.init(); err != nil {
+		return err
+	}
+
+	// We have to make this check because the ListWatch that we use in
+	// WatchServices requires Client functions that aren't in the interface
+	// for some reason.
+	if _, ok := s.kubeClient.(*clientset.Clientset); !ok { // TODO: quinton: Should we use a generic kubeClient here, or a more specific ubeClient?
+		return fmt.Errorf("ServiceController only works with real Client objects, but was passed something else satisfying the clientset.Interface.")
+	}
+
+	// Get the currently existing set of Federated services and then all future creates
+	// and updates of these services.
+	// A delta compressor is needed for the DeltaFIFO queue because we only ever
+	// care about the most recent state.
+	serviceQueue := cache.NewDeltaFIFO(
+		cache.MetaNamespaceKeyFunc,
+		cache.DeltaCompressorFunc(func(d cache.Deltas) cache.Deltas {
+			if len(d) == 0 {
+				return d
+			}
+			return cache.Deltas{*d.Newest()}
+		}),
+		s.cache,
+	)
+	lw := cache.NewListWatchFromClient(s.kubeClient.(*clientset.Clientset).CoreClient, "services", api.NamespaceAll, fields.Everything())
+	cache.NewReflector(lw, &api.Service{}, serviceQueue, serviceSyncPeriod).Run()
```

Comment:
I took a look on all related discussion for the rewriting of service controller, the discussion is regarding to how to watch the service deltas, and handle them in 10 go routines, and I agree that part need to be refactored to apply the best pattern like other controllers.

My comment here is not for the service delta, the logic is special for service controller, and there is no extra go routine to handle the logic. This function is watching the logic of k8s cluster changes, it obverse the changes and apply all services to all latest clusters. We can have more concrete discussions related to this part when dns api is ready.